### PR TITLE
Bubble Sort Tutorial

### PR DESCRIPTION
**Pull request by RadhikaChhabra17**
## Change Summary

Tutorial Writeup - Bubble Sort #93 .

## Checklist

**Ideally all of them should be checked. If you haven't fulfilled the below requirements or even delete the entire checklist, your PR won't be reviewed.**

### General 

- [x] Make sure you've read [CONTRIBUTING.md](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md) 
- [x] Make sure you've followed the template format
- [x] Make sure you've started the app locally and verified all the content and all links (if applicable) must be accessible correctly
- [x] Make sure you've applied LaTex for variables, formulas and time / space complexity instead of using backticks
- [x] Make sure you've formatted your code well (Code Block is rendered using `<Tabs/>`)
      Below is a C++ example. You can include multiple langs using `<TabItem/>` if necessary
      
````
<Tabs>

<TabItem value="cpp" label="C++">
<SolutionAuthor name="@YOUR_ALIAS"/>

```cpp
// Your code goes here
// Prefer K&R Coding style
// See https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr for example
```

</TabItem>

</Tabs>
````
- [x] Make sure you've explained your code well (write comments above each line - not inline)
- [x] Make sure you've written your explanation well and it is easy to understand for beginners
- [x] Make sure you've included Time & Space Complexity Analysis
- [x] Make sure you've included a meaningful approach name for your solution. Don't leave it blank. e.g. `## Approach 1: Two Pointers`.

### For Tutorial Articles

- [x] Make sure you've explained your topic using 1 - 3 LC problems rather than using external problems. Solutions must be taken with explanations as well.
- [x] Make sure you've provided suggested problems at the end using the given format. See [here](https://raw.githubusercontent.com/wingkwong/leetcode-the-hard-way/main/tutorials/math/number-theory/binary-exponentiation.md) as an example. If the target solution is not available, leave `solutionLink` blank.
- [x] Make sure you've given credits / references if you use external resources. For an image, give credit under it. Otherwise, add a new section called References at the end (after Suggested Problems).

### CODE REVIEW

**Review by wingkwong — CHANGES_REQUESTED** [review]
Please look at the checklist first as well as other tutorials for reference.

**Comment by RadhikaChhabra17** [comment]
Can you please review it and let me know if any changes are required.

**Review by wingkwong — CHANGES_REQUESTED** [review]
Please look at the checklist first as well as other tutorials for reference.

**Comment by RadhikaChhabra17** [comment]
Hi, it would be helpful if you can be more specific about change.

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 1]
```diff
@@ -1,14 +1,156 @@
...
+- leetcode
+
+- tutorial
+
+- sorting
+
+- algorithm
+
+- bubble sort
+
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:20]: unexpected line breaks

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 2]
```diff
@@ -1,14 +1,156 @@
...
+1. Make two nested loop, the 1st loop would be for number of pass the algorithm woud run, total n-1 passes and second would be for comparison in that pass. In each pass, one element is sorted (largest or smallest, as required) and placed in correct position and rest are compared in further passes.
+2. If one element is bigger(or smaller in decreasing order case) than its next element, then both should be swapped. 
+
+
+  
+
+Consider the example of unsorted list and see how the algorithm works.
+
+  
+
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:49]: there's lots of extra line breaks

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 3]
```diff
@@ -1,14 +1,156 @@
...
+	i = 0 -> 3,4,2,5,6
+	i = 1 -> 3,2,4,5,6
+	i = 2 -> 3,2,4,5,6
+
+3rd pass = 
+	i = 0 -> 2,3,4,5,6
+	i = 1 -> 2,3,4,5,6
+
+4th pass =
+	i = 0 -> 2,3,4,5,6
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:72]: put it outside and present it better

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 4]
```diff
@@ -1,14 +1,156 @@
...
+		for(int j=0; j<n-i-1; j++) {
+			if(arr[j] > arr[j+1]) {
+				int temp = arr[j];
+				arr[j] = arr[j+1];
+				arr[j+1] = temp;
+				check = true;
+			}
+		}
+	}
+}
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:90]: code should be in <Tabs/> .. just like what you did in solutions.

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 5]
```diff
@@ -1,14 +1,156 @@
...
+### Time and Space Complexity
+
+Time Complexity :
+For first iteration, loop will run (n-1) times, for second it will run (n-2) times and so on.
+So, Worst Case Time Complexity = (n-1) + (n-2) + (n-3) + ... + 1 = O(n^2)
+If there is no change in any iterations, next iterations will not occurs, this reduces the time complexity.
+
+Space Complexity :
+Since there is no extra space, Space Complexity = O(1)
+This shows that it is an inline sorting.
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:102]: 1. these are packed together. Verify it in your local by running the app. (Not looking at markdown editor)
2. as stated in checklist, you need to apply LaTex on variables / formula / time & space complexity

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 6]
```diff
@@ -1,14 +1,156 @@
...
+>
+>Return the array formed after rearranging the values of nums.
+
+*Approach* : Run two bubble sort. One for even indices with non decreasing order and one for odd indices with non increasing order.
+
+```c++
+class Solution {
+public:
+    vector<int> sortEvenOdd(vector<int>& nums) {
+        //Working for even indices
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:118]: // work for even indices

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 7]
```diff
@@ -1,14 +1,156 @@
...
+        for (int i = 0; i < nums.size(); i += 2){
+            for (int j = i+2; j < nums.size(); j += 2){
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:120]: i + 2

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 8]
```diff
@@ -1,14 +1,156 @@
...
+                //sorting in non-decreasing order
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:121]: // sort in non-decreasing order

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 9]
```diff
@@ -1,14 +1,156 @@
...
+                if (nums[i] > nums[j]){
+                    swap(nums[i], nums[j]);
+                }
+            }
+        }
+        //Working for odd indicies
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:127]: // work for odd indicies

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 10]
```diff
@@ -1,14 +1,156 @@
...
+        for (int i = 1; i < nums.size(); i += 2){
+           for (int j = i+2; j < nums.size(); j += 2){
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:129]: i + 2

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 11]
```diff
@@ -1,14 +1,156 @@
...
+               //sorting in non-increasing order
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:130]: // sort in non-increasing order

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 12]
```diff
@@ -1,14 +1,156 @@
...
+
+<TutorialAuthors  names="@RadhikaChhabra17"/>
+
+  
+
+## Overview
+
+  
+
+Bubble Sort is a sorting algorithm which compares the adjacent elements and swap their positions if they are placed in wrong order. At max, we need to compare adjacent elements for (n-1) iterations where n is the size of array to be sorted. At the end of each iteration, larger(or smaller, as required) value is sorted and placed at correct positions.
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:33]: 1. Apply LaTex on `n-1`, `n`. e.g. `$n - 1$` and `$n$`
2. add a space after the word `larger`

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 13]
```diff
@@ -1,14 +1,156 @@
...
+
+4th pass =
+	i = 0 -> 2,3,4,5,6
+
+Code :
+
+void bubblesort(vector<int> &arr) {
+	int n = arr.size();
+	bool check = true;
+	for(int i=0; i<n-1 && check; i++) {
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:79]: have you explained why we need `check` here?

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 14]
```diff
@@ -1,14 +1,156 @@
...
+		check = false;
+		for(int j=0; j<n-i-1; j++) {
+			if(arr[j] > arr[j+1]) {
+				int temp = arr[j];
+				arr[j] = arr[j+1];
+				arr[j+1] = temp;
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:85]: why not just `swap(arr[j], arr[j + 1])` ?

**Review by wingkwong — CHANGES_REQUESTED** [review]

**Comment by RadhikaChhabra17** [comment]
I have made the required changes. Please look into it and let me know if any change is required.

File: solutions/0000-0099/0075-sort-colors-medium.md [hunk 15]
```diff
@@ -72,4 +72,37 @@ class Solution {
...
+    
+Time Complexity: $O(n^2)$, where $n$ - elements in array
+
+Space complexity: $O(1)$
+
+<Tabs>
+<TabItem value="c++" label="C++">
+<SolutionAuthor name="@RadhikaChhabra17"/>
+
+```c++
```
Comment (wingkwong) [inline solutions/0000-0099/0075-sort-colors-medium.md:89]: same. K&R coding style & spacing.

File: solutions/0100-0199/0148-sort-list-medium.md [hunk 16]
```diff
@@ -129,4 +129,39 @@ class Solution {
...
+
+Time Complexity: $O(n^2)$, where $n$ - # of nodes in the list
+
+Space complexity: $O(1)$
+ <Tabs>
+<TabItem value="c++" label="C++">
+<SolutionAuthor name="@RadhikaChhabra17"/>
+
+```c++
+class Solution {
```
Comment (wingkwong) [inline solutions/0100-0199/0148-sort-list-medium.md:146]: same. K&R coding style & spacing. comments start with lower case

File: solutions/0100-0199/0148-sort-list-medium.md [hunk 17]
```diff
@@ -129,4 +129,39 @@ class Solution {
 ```
 
 </TabItem>
-</Tabs>
\ No newline at end of file
+</Tabs>
+    
+ # Approach 2: Bubble Sort
```
Comment (wingkwong) [inline solutions/0100-0199/0148-sort-list-medium.md:134]: remove extra space

File: solutions/0100-0199/0148-sort-list-medium.md [hunk 18]
```diff
@@ -129,4 +129,39 @@ class Solution {
...
+
+We can use the same concept of bubble sort and code it in context of linked list.
+
+Time Complexity: $O(n^2)$, where $n$ - # of nodes in the list
+
+Space complexity: $O(1)$
+ <Tabs>
```
Comment (wingkwong) [inline solutions/0100-0199/0148-sort-list-medium.md:141]: remove extra space and add a line break

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 19]
```diff
@@ -1,14 +1,149 @@
...
+Consider the example of unsorted list and see how the algorithm works.
+
+arr = {3,6,4,2,5}: 
+
+1st pass =>
+    
+	index = 0 -> 3,6,4,2,5 
+	index = 1 -> 3,4,6,2,5
+	index = 2 -> 3,4,2,6,5
+    index = 3 -> 3,4,2,5,6
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:34]: not aligned

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 20]
```diff
@@ -1,14 +1,149 @@
...
+
+The syntax that we are using is `cpp`, please refer to your own language of preference if needed.
+
+Algorithm :
+1. Make two nested loop, the 1st loop would be for number of pass the algorithm woud run, total $n-1$ passes and second would be for comparison in that pass. In each pass, one element is sorted (largest or smallest, as required) and placed in correct position and rest are compared in further passes.
+2. If one element is bigger (or smaller in decreasing order case) than its next element, then both should be swapped. 
+
+Consider the example of unsorted list and see how the algorithm works.
+
+arr = {3,6,4,2,5}:
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:27]: `$arr = [3, 6, 4, 2, 5]$`

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 21]
```diff
@@ -1,14 +1,149 @@
...
 ---
 
-<TutorialAuthors names="@TBC"/>
+<TutorialAuthors  names="@RadhikaChhabra17"/>
+
+## Overview
+
+Bubble Sort is a sorting algorithm which compares the adjacent elements and swap their positions if they are placed in wrong order. At max, we need to compare adjacent elements for $$(n - 1)$$ iterations where $n$ is the size of array to be sorted. At the end of each iteration, larger (or smaller, as required) value is sorted and placed at correct positions.
+
+The syntax that we are using is `cpp`, please refer to your own language of preference if needed.
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:19]: no need to mention this line.

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 22]
```diff
@@ -1,14 +1,149 @@
...
+3rd pass => 
+
+	index = 0 -> 2,3,4,5,6
+	index = 1 -> 2,3,4,5,6
+
+4th pass =>
+
+	index = 0 -> 2,3,4,5,6  
+	
+We can use a variable *check* to see if there is swap in one pass or not. If there is no swapping in one pass, they we don't have to check for other pass.
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:51]: prefer to show the standard bubble sort code first. then mention how to optimize it using `check`. besides, this line is not clean.

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 23]
```diff
@@ -1,14 +1,149 @@
...
+
+ <Tabs>
+<TabItem value="c++" label="C++">
+<SolutionAuthor name="@RadhikaChhabra17"/>
+
+```c++
+void bubblesort(vector<int> &arr) {
+	int n = arr.size();
+	bool check = true;
+	for(int i = 0; i < n - 1 && check; i++) {
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:61]: there should be a space after `for`. please check all.

`for (int i = 0; i < n - 1 && check; i++) {`

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 24]
```diff
@@ -1,14 +1,149 @@
...
+		check = false;
+		for(int j = 0; j < n - i - 1; j++) {
+			if(arr[j] > arr[j + 1]) {
+                swap(arr[j], arr[j + 1]);
+				check = true;
+			}
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:67]: 1. poor indentation
2. add a space after `if`

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 25]
```diff
@@ -1,14 +1,149 @@
...
+
+<Tabs>
+<TabItem value="c++" label="C++">
+<SolutionAuthor name="@RadhikaChhabra17"/>
+
+```c++
+class Solution {
+public:
+    vector<int> sortEvenOdd(vector<int>& nums) {
+        //For even indices
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:109]: i expected you to copy the line in my previous comments...

1. lower case
2. add a space after `//`

apply to all.

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 26]
```diff
@@ -1,14 +1,149 @@
...
+</TabItem>
+</Tabs>
+
+### Time and Space Complexity
+
+Time Complexity :
+
+For first iteration, loop will run ($n-1$) times, for second it will run ($n-2$) times and so on.
+
+So, Worst Case Time Complexity = $$(n - 1)$$ + $$(n - 2)$$ + $$(n - 3)$$ + ... + 1 = $$O(n^2)$$
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:81]: Too tedious. Use

```
$(n - 1) + (n - 2) + (n - 3) + ... + 1 = O(n ^ 2)$
```

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 27]
```diff
@@ -1,14 +1,149 @@
...
+Use of variable If there is no change in any iterations, next iterations will not occurs, this reduces the time complexity.
+
+Space Complexity :
+Since there is no extra space, Space Complexity = $$O(1)$$
+This shows that it is an inline sorting.
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:86]: these will be rendered in the same line in the app. please verify it locally.

File: tutorials/basic-topics/sorting/bubble-sort.md [hunk 28]
```diff
@@ -1,14 +1,149 @@
...
+	index = 2 -> 3,2,4,5,6
+
+3rd pass => 
+
+	index = 0 -> 2,3,4,5,6
+	index = 1 -> 2,3,4,5,6
+
+4th pass =>
+
+	index = 0 -> 2,3,4,5,6
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bubble-sort.md:49]: not intuitive at all

**Review by wingkwong — CHANGES_REQUESTED** [review]

**Review by wingkwong — APPROVED** [review]
Cleaned up a bit.

**Comment by wingkwong** [comment]
@RadhikaChhabra17 Your solution gives TLE in 0148 - Sort List (Medium). I removed it. Besides, please take a look the commits that I've made above. You will have ideas of what you've missed. I'll merge this. Thanks.

**Pull request merged**